nmostesting/suites/BCP0060101Test.py

```diff
@@ -527,6 +527,162 @@ def test_04(self, test):
+            # sampling corresponds to Flow 'components' so there isn't a Flow schema to use
+            "urn:x-nmos:cap:format:color_sampling": [
+                # Red-Green-Blue-Alpha
+                "RGBA",
+                # Red-Green-Blue
+                "RGB",
+                # Non-constant luminance YCbCr
+                "YCbCr-4:4:4",
+                "YCbCr-4:2:2",
+                "YCbCr-4:2:0",
+                "YCbCr-4:1:1",
+                # Constant luminance YCbCr
+                "CLYCbCr-4:4:4",
+                "CLYCbCr-4:2:2",
+                "CLYCbCr-4:2:0",
+                # Constant intensity ICtCp
+                "ICtCp-4:4:4",
+                "ICtCp-4:2:2",
+                "ICtCp-4:2:0",
+                # XYZ
+                "XYZ",
+                # Key signal represented as a single component
+                "KEY",
+                # Sampling signaled by the payload
+                "UNSPECIFIED"
+            ],
+            "urn:x-nmos:cap:transport:packet_transmission_mode": sender_properties["packet_transmission_mode"]["enum"],
+            "urn:x-nmos:cap:transport:st2110_21_sender_type": sender_properties["st2110_21_sender_type"]["enum"],
+        }
+
+        try:
+            jxsv_receivers = [receiver for receiver in self.is04_resources["receivers"]
+                              if "media_types" in receiver["caps"]
+                              and "video/jxsv" in receiver["caps"]["media_types"]]
+
+            for receiver in jxsv_receivers:
+                # check required attributes are present
+                if "constraint_sets" not in receiver["caps"]:
+                    # FAIL reported by test_05
+                    continue
+
+                # exclude constraint sets for other media types
+                jxsv_constraint_sets = [constraint_set for constraint_set in receiver["caps"]["constraint_sets"]
+                                        if media_type_constraint not in constraint_set
+                                        or ("enum" in constraint_set[media_type_constraint]
+                                            and "video/jxsv" in constraint_set[media_type_constraint]["enum"])]
+
+                if len(jxsv_constraint_sets) == 0:
+                    # FAIL reported by test_05
+                    continue
+
+                # check recommended attributes are present
+                for constraint_set in jxsv_constraint_sets:
+                    for constraint, enum_values in enum_constraints.items():
+                        if constraint in constraint_set and "enum" in constraint_set[constraint]:
+                            for enum_value in constraint_set[constraint]["enum"]:
+                                if enum_value not in enum_values:
+                                    return test.FAIL("Receiver {} uses an invalid value for '{}': {}"
+                                                     .format(receiver["id"], constraint, enum_value))
```

Comment:
Could these nested loops be replaced with itertools? Something like:

```
import itertools

for constraint_set, constraint in itertools.product(jxsv_constraint_sets, enum_constraints):
    if constraint in constraint_set and "enum" in constraint_set[constraint]:
        ...
```

although it only gets rid of one level of for loop.  Although there may be a way of combining all three loops perhaps?

Reply:
Yeah, maybe, I'll try! I see our very first use of `itertools` is in `BCP0060102Test`. 😉

Reply:
I'm not sure I'm finding this much more readable... one fewer indent but to meet the line length, the same number of lines.

```
                for constraint_set, (constraint, enum_values)
                        in product(jxsv_constraint_sets, enum_constraints.items()):
                    ...
```

I'm going to stick with the nested `for` loops in this case, but I'm sure there are more uses for `itertools` in our test code.